client/src/main/java/org/apache/rocketmq/client/impl/consumer/DefaultMQPushConsumerImpl.java

```diff
@@ -218,25 +219,49 @@ public void pullMessage(final PullRequest pullRequest) {
             return;
         }
 
+        //flow control for topic
+        if (this.defaultMQPushConsumer.getPullThresholdForTopic() != Integer.MAX_VALUE) {
+            Map<MessageQueue, ProcessQueue> allProcessQMap = this.getRebalanceImpl().getProcessQueueTable();
+            Iterator<Entry<MessageQueue, ProcessQueue>> it = allProcessQMap.entrySet().iterator();
+            long sizeOfAllQueue = 0;
+            //pick the relative process queues and calculate size
+            while (it.hasNext()) {
+                Entry<MessageQueue, ProcessQueue> entry = it.next();
+                if (pullRequest.getMessageQueue().getTopic().equals(entry.getKey().getTopic())) {
+                    sizeOfAllQueue += entry.getValue().getMsgCount().get();
+                }
+            }
+            if (sizeOfAllQueue > this.defaultMQPushConsumer.getPullThresholdForTopic()) {
+                this.executePullRequestLater(pullRequest, PULL_TIME_DELAY_MILLS_WHEN_FLOW_CONTROL);
+                if ((topicFlowControlTimes++ % 1000) == 0) {
+                    log.warn(
+                        "the consumer message topic buffer is full, so do flow control, minOffset={}, maxOffset={}, sizeOfAllQueue={}, pullRequest={}, flowControlTimes={}",
```

Comment:
`so do flow control` is very ambiguous. Let's have a more concrete advice on how to cope with the situation.